### PR TITLE
abi: return number of bytes read from abi.Decode

### PR DESCRIPTION
**Pull request by ryandotsmith**
It is not uncommon for ethereum calldata to include an abi encoded payload with extra bytes appended. In order to seperate the abi encoded data from the extra bytes, Decode will return the number of bytes read so that callers can safely extract the extra bytes.

This commit adds a test to ensure that Decode will work with input that contains extra bytes.

### CODE REVIEW

File: abi/abi.go [hunk 1]
```diff
@@ -150,46 +150,58 @@ func (item *Item) Done() {
 // Decodes ABI encoded bytes into an [Item] according to
 // the 'schema' defined by t. For example:
 //	Decode(b, schema.Tuple(schema.Dynamic(), schema.Static()))
-func Decode(input []byte, t schema.Type) *Item {
+// Returns the item and the number of bytes read from input
+func Decode(input []byte, t schema.Type) (int, *Item) {
```
Comment (worm-emoji) [inline abi/abi.go:154]: bit of drive-by feedback: I feel like it makes more sense from a Go standpoint to return (*Item, int). Felt weird writing code against this that had it the other way around.

Comment (ryandotsmith) [reply abi/abi.go:154]: Is there Go code you cant point to where n is not first? Here are some examples of where it is first:

- https://pkg.go.dev/bytes#Buffer.Read
- https://pkg.go.dev/io#Reader

Comment (worm-emoji) [reply abi/abi.go:154]: Yeah, totally hear you there. My rationale is that I feel like `*Item` is the primary / most relevant value that decode is returning, not size.

Even in one of the packages you shared, `ReadRune` is similar to what I mean:

* https://pkg.go.dev/bytes#Buffer.ReadRune

Comment (worm-emoji) [reply abi/abi.go:154]: I also think that `Buffer.Read` is potentially a bad example because it mutates the pointer its on. If it was a function that actually returned a value (e.g. the buffer or similar), I bet that would be the first returned parameter, not the size.

Comment (worm-emoji) [reply abi/abi.go:154]: Feel free to disagree, just wanted to make my case :)
